internal/agentcfg/elasticsearch.go

```diff
@@ -0,0 +1,259 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package agentcfg
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/monitoring"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+const ElasticsearchIndexName = ".apm-agent-configuration"
+
+const (
+	// ErrInfrastructureNotReady is returned when a fetch request comes in while
+	// the infrastructure is not ready to serve the request.
+	// This may happen when the local cache is not initialized and no fallback fetcher is configured.
+	ErrInfrastructureNotReady = "agentcfg infrastructure is not ready"
+
+	// ErrNoValidElasticsearchConfig is an error where the server is
+	// not properly configured to fetch agent configuration.
+	ErrNoValidElasticsearchConfig = "no valid elasticsearch config to fetch agent config"
+)
+
+const refreshCacheTimeout = 5 * time.Second
+
+type ElasticsearchFetcher struct {
+	client          *elasticsearch.Client
+	cacheDuration   time.Duration
+	fallbackFetcher Fetcher
+
+	mu    sync.RWMutex
+	last  time.Time
+	cache []AgentConfig
+
+	searchSize int
+
+	invalidESCfg     atomic.Bool
+	cacheInitialized atomic.Bool
+
+	logger *logp.Logger
+
+	metrics fetcherMetrics
+}
+
+type fetcherMetrics struct {
+	fetchES, fetchFallback, fetchFallbackUnavailable,
+	cacheRefreshSuccesses, cacheRefreshFailures,
+	cacheEntriesCount atomic.Int64
+}
+
+func NewElasticsearchFetcher(client *elasticsearch.Client, cacheDuration time.Duration, fetcher Fetcher) *ElasticsearchFetcher {
+	logger := logp.NewLogger("agentcfg")
+	return &ElasticsearchFetcher{
+		client:          client,
+		cacheDuration:   cacheDuration,
+		fallbackFetcher: fetcher,
+		searchSize:      100,
+		logger:          logger,
+	}
+}
+
+// Fetch finds a matching agent config based on the received query.
+func (f *ElasticsearchFetcher) Fetch(ctx context.Context, query Query) (Result, error) {
+	if f.cacheInitialized.Load() {
+		// Happy path: serve fetch requests using an initialized cache.
+		f.mu.RLock()
+		defer f.mu.RUnlock()
+		f.metrics.fetchES.Add(1)
+		return matchAgentConfig(query, f.cache), nil
+	}
+
+	if f.fallbackFetcher != nil {
+		f.metrics.fetchFallback.Add(1)
+		return f.fallbackFetcher.Fetch(ctx, query)
+	}
+
+	if f.invalidESCfg.Load() {
+		f.metrics.fetchFallbackUnavailable.Add(1)
```

Comment:
Do we want a separate metric for this "no valid elasticsearch config" case?

Reply:
yes that sounds reasonable

Reply:
Added `agentcfg.elasticsearch.fetch.invalid`